File: code/library/src/main/java/org/fbme/lib/iec61499/stringify/FBInterfacePrinter.java
package org.fbme.lib.iec61499.stringify;


import org.fbme.lib.iec61499.declarations.EventAssociation;
import org.fbme.lib.iec61499.declarations.EventDeclaration;
import org.fbme.lib.iec61499.declarations.FBInterfaceDeclaration;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class FBInterfacePrinter extends PrinterElementBase<FBInterfaceDeclaration> {

    private final boolean mySubApp;

    public FBInterfacePrinter(FBInterfaceDeclaration declaration, boolean subApp) {
        super(declaration, "InterfaceList");
        mySubApp = subApp;
    }

    @Override
    protected void printElementBody(Element element) {
        if (mySubApp) {
            printInterfaceWithSubAppEvents(element);
        } else {
            printInterfaceWithEvents(element);
        }
    }

    private void printInterfaceWithEvents(Element element) {
        addNullableContent(element, printEvents("EventInputs", myElement.getInputEvents()));
        addNullableContent(element, printEvents("EventOutputs", myElement.getOutputEvents()));
        addNullableContent(element, ParameterDeclarationPrinter.printAll("InputVars", myElement.getInputParameters()));
        addNullableContent(element, ParameterDeclarationPrinter.printAll("OutputVars", myElement.getOutputParameters()));
    }

    private void printInterfaceWithSubAppEvents(Element element) {
        addNullableContent(element, printSubAppEvents("SubAppEventInputs", myElement.getInputEvents()));
        addNullableContent(element, printSubAppEvents("SubAppEventOutputs", myElement.getOutputEvents()));
        addNullableContent(element, ParameterDeclarationPrinter.printAll("InputVars", myElement.getInputParameters()));
        addNullableContent(element, ParameterDeclarationPrinter.printAll("OutputVars", myElement.getOutputParameters()));
    }

    private Element printEvents(String collectionTag, List<EventDeclaration> events) {
        if (events.isEmpty()) {
            return null;
        }
        Element element = new Element(collectionTag);
        for (EventDeclaration eventDeclaration : events) {
            element.addContent(EventPrinter.event(eventDeclaration).print());
        }
        return element;
    }

    private Element printSubAppEvents(String collectionTag, List<EventDeclaration> events) {
        if (events.isEmpty()) {
            return null;
        }
        Element element = new Element(collectionTag);
        for (EventDeclaration eventDeclaration : events) {
            element.addContent(EventPrinter.subAppEvent(eventDeclaration).print());
        }
        return element;
    }

    private static class EventPrinter extends DeclarationPrinterBase<EventDeclaration> {

        private EventPrinter(String tag, EventDeclaration eventDeclaration) {
            super(eventDeclaration, tag);
        }

        @Override
        protected void printDeclarationBody(@NotNull Element element) {
            for (EventAssociation association : myElement.getAssociations()) {
                Element with = new Element("With");
                with.setAttribute("Var", association.getParameterReference().getPresentation());
                element.addContent(with);
            }
        }

        private static EventPrinter event(EventDeclaration eventDeclaration) {
            return new EventPrinter("Event", eventDeclaration);
        }

        private static EventPrinter subAppEvent(EventDeclaration eventDeclaration) {
            return new EventPrinter("SubAppEvent", eventDeclaration);
        }
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/stringify/FBInterfacePrinter.kt
package org.fbme.lib.iec61499.stringify

import org.fbme.lib.iec61499.declarations.EventDeclaration
import org.fbme.lib.iec61499.declarations.FBInterfaceDeclaration
import org.jdom.Element

class FBInterfacePrinter(declaration: FBInterfaceDeclaration, private val mySubApp: Boolean) :
    PrinterElementBase<FBInterfaceDeclaration>(declaration, if (mySubApp) "SubAppInterfaceList" else "InterfaceList") {
    override fun printElementBody(element: Element) {
        if (mySubApp) {
            printInterfaceWithSubAppEvents(element)
        } else {
            printInterfaceWithEvents(element)
        }
    }

    private fun printInterfaceWithEvents(element: Element) {
        addNullableContent(element, printEvents("EventInputs", this.element.inputEvents))
        addNullableContent(element, printEvents("EventOutputs", this.element.outputEvents))
        addNullableContent(element, ParameterDeclarationPrinter.printAll("InputVars", this.element.inputParameters))
        addNullableContent(element, ParameterDeclarationPrinter.printAll("OutputVars", this.element.outputParameters))
    }

    private fun printInterfaceWithSubAppEvents(element: Element) {
        addNullableContent(element, printSubAppEvents("SubAppEventInputs", this.element.inputEvents))
        addNullableContent(element, printSubAppEvents("SubAppEventOutputs", this.element.outputEvents))
        addNullableContent(element, ParameterDeclarationPrinter.printAll("InputVars", this.element.inputParameters))
        addNullableContent(element, ParameterDeclarationPrinter.printAll("OutputVars", this.element.outputParameters))
    }

    private fun printEvents(collectionTag: String, events: List<EventDeclaration?>): Element? {
        if (events.isEmpty()) {
            return null
        }
        val element = Element(collectionTag)
        for (eventDeclaration in events) {
            element.addContent(EventPrinter.event(eventDeclaration).print())
        }
        return element
    }

    private fun printSubAppEvents(collectionTag: String, events: List<EventDeclaration?>): Element? {
        if (events.isEmpty()) {
            return null
        }
        val element = Element(collectionTag)
        for (eventDeclaration in events) {
            element.addContent(EventPrinter.subAppEvent(eventDeclaration).print())
        }
        return element
    }

    private class EventPrinter private constructor(tag: String, eventDeclaration: EventDeclaration) :
        DeclarationPrinterBase<EventDeclaration>(eventDeclaration, tag) {
        override fun printDeclarationBody(element: Element) {
            for (association in this.element.associations) {
                val with = Element("With")
                with.setAttribute("Var", association.parameterReference.presentation)
                element.addContent(with)
            }
        }

        companion object {
            fun event(eventDeclaration: EventDeclaration?): EventPrinter {
                return EventPrinter("Event", eventDeclaration!!)
            }

            fun subAppEvent(eventDeclaration: EventDeclaration?): EventPrinter {
                return EventPrinter("SubAppEvent", eventDeclaration!!)
            }
        }
    }
}
